File: check_abuseipdb.py
#! /usr/bin/env python3

import sys
import ipaddress
import urllib.request
import simplejson
from optparse import OptionParser
from AbuseIpDbCheckOptions import AbuseIpDbCheckOptions

VERSION='0.1.0'
USER_AGENT='abuseipdb_checkscript/%s (Python3/urllib; Github: webfoersterei)' % (VERSION)
TIMEOUT=5
BASEURL="https://www.abuseipdb.com/check/{}"

# From https://www.abuseipdb.com/categories
CATEGORY_NAMES={3: 'Fraud Orders', 4: 'DDoS Attack',
                9: 'Open Proxy', 10: 'Web Spam',
                11: 'Email Spam', 14: 'Port Scan',
                18: 'Brute-Force',19: 'Bad Web Bot',
                20: 'Exploited Host',21: 'Web App Attack',
                22: 'SSH', 23: 'IoT Targeted'}

# define exit codes
EXIT_OK = 0
EXIT_WARN = 1
EXIT_CRIT = 2
EXIT_UNKNOWN = 3

def queryEntriesFromApi(checkOptions: AbuseIpDbCheckOptions) -> object:
    ''' Will get entries from API and return an python object with all entries '''
    apiurl = BASEURL + '/json?key={}&days={}'    
    requestUrl = apiurl.format(checkOptions.hostaddress, checkOptions.apiKey, checkOptions.daysToQuery)
    headers = {'User-Agent': USER_AGENT}
    request = urllib.request.Request(requestUrl, None, headers)
    
    response = urllib.request.urlopen(request, timeout=TIMEOUT).read()

    return simplejson.loads(response)

def analyseEntries(checkOptions: AbuseIpDbCheckOptions, entries: list) -> int:
    ''' Will print some information and return the exit code '''
    if(len(entries) == 0):
        print("OK - No entries found")
        sys.exit(EXIT_OK)
    elif(len(entries) >= checkOptions.criticalThreshold):
        print('CRITICAL', end=' - ')
        exitCode = EXIT_CRIT
    elif(len(entries) >= checkOptions.warningThreshold):
        print('WARN', end=' - ')
        exitCode = EXIT_WARN
    else:
        # EntryCount is not 0 but also no threshold was reached
        print('OK', end=' - ')
        exitCode = EXIT_OK

    categories = set()

    for entry in entries:
        if len(entry['category']) > 0:
            for cat in entry['category']:
                categories.add(cat)
    
    categoryNames = []
    for category in categories:
        if category in CATEGORY_NAMES:
            categoryNames.append(CATEGORY_NAMES[category])

    print("Reported {0}x (last {1}d) for: {2}".format(len(entries), checkOptions.daysToQuery, ', '.join(categoryNames)))
    return exitCode

def normalizeOptions(hostaddress, apiKey, warningThreashold, criticalThreshold, daysToQuery):
    return AbuseIpDbCheckOptions(hostaddress, apiKey, warningThreashold, criticalThreshold, daysToQuery)

def main():
    parser = OptionParser("usage: %prog -H <IP address> -K <API key>")
    parser.add_option("-H","--hostaddress", dest="host", help="Specify the IP address you want to check")
    parser.add_option("-V","--version", action="store_true", dest="version", help="This option show the current version number of the program and exit")
    parser.add_option("-K", "--key", dest="key", help="API-key for abuseipdb.com")
    parser.add_option("-w", "--warn", dest="warningCount", default=1, type=int, help="Threshold for entries so that the return is a WARN")
    parser.add_option("-c", "--crit", dest="criticalCount", default=3, type=int, help="Threshold for entries so that the return is a CRIT")
    parser.add_option("-d", "--days", dest="days", default=14, type=int, help="Maximum age of reports to take into consideration")

    (opts, _) = parser.parse_args()

    if opts.version:
        print("check_abuseipdb.py %s"%VERSION)
        sys.exit()

    exitCode = EXIT_UNKNOWN
    checkOptions = normalizeOptions(opts.host, opts.key, opts.warningCount, opts.criticalCount, opts.days)
    
    try:
        entries = queryEntriesFromApi(checkOptions)
        
        if(type(entries) == dict):
            entries = [entries]
        
        exitCode = analyseEntries(checkOptions, entries)
    except Exception as ex:
        print('UNKNOWN - Problem querying API: {0}'.format(ex))
        sys.exit(EXIT_UNKNOWN)

    sys.exit(exitCode)
 
if __name__ == '__main__':
    main()

File: check_abuseipdb.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"
)

var version string
var compileDate string

const (
	ExitCodeOk = iota
	ExitCodeWarning
	ExitCodeCritical
	ExitCodeUnknown
)

var AbuseIpDbCategories = map[int]string{
	3: "Fraud Orders", 4: "DDoS Attack",
	5: "FTP Brute-Force", 6: "Ping of Death",
	7: "Phishing", 8: "Fraud VoIP",
	9: "Open Proxy", 10: "Web Spam",
	11: "Email Spam", 12: "Blog Spam",
	13: "VPN IP", 14: "Port Scan",
	15: "Hacking", 16: "SQL Injection",
	17: "Spoofing", 18: "Brute-Force",
	19: "Bad Web Bot", 20: "Exploited Host",
	21: "Web App Attack", 22: "SSH", 23: "IoT Targeted",
}

type AbuseIpDbCheckResponse struct {
	Data AbuseIpDbCheckData `json:"data"`
}

type AbuseIpDbCheckData struct {
	IpAddress            string                 `json:"ipAddress"`
	IsPublic             bool                   `json:"isPublic"`
	IpVersion            int                    `json:"ipVersion"`
	IsWhitelisted        bool                   `json:"isWhitelisted"`
	AbuseConfidenceScore int                    `json:"abuseConfidenceScore"`
	CountryCode          string                 `json:"countryCode"`
	UsageType            string                 `json:"usageType"`
	Isp                  string                 `json:"isp"`
	Domain               string                 `json:"domain"`
	TotalReports         int                    `json:"totalReports"`
	NumDistinctUsers     int                    `json:"numDistinctUsers"`
	LastReportedAt       string                 `json:"lastReportedAt"`
	Reports              []AbuseIpDbCheckReport `json:"reports"`
}

type AbuseIpDbCheckReport struct {
	ReportedAt          string `json:"reportedAt"`
	Comment             string `json:"comment"`
	Categories          []int  `json:"categories"`
	ReporterId          int    `json:"reporterId"`
	ReporterCountryCode string `json:"reporterCountryCode"`
	ReporterCountryName string `json:"reporterCountryName"`
}

func buildStatusMessage(checkData AbuseIpDbCheckData) string {
	var reasonStr string

	if len(checkData.Reports) > 0 {
		var reasons = make(map[int]string) // Slice of *unique* reason strings
		for _, report := range checkData.Reports {
			for _, categoryId := range report.Categories {
				reasons[categoryId] = AbuseIpDbCategories[categoryId]
			}
		}

		for _, reason := range reasons {
			reasonStr += reason + ", "
		}
	}
	return strings.TrimRight(fmt.Sprintf("Found %d entries for %s from %d users (Abuse Probability: %d%%) %s",
		checkData.TotalReports, checkData.IpAddress, checkData.NumDistinctUsers, checkData.AbuseConfidenceScore, reasonStr), ", ")
}

func main() {
	address := flag.String("host", "", "Host to check")
	apiKey := flag.String("key", "", "abuseipdb APIv2 key")
	daysToCheck := flag.Int("days", 14, "Timespan to check in days")
	warnCount := flag.Int("warn", 1, "Minimum reports to return a WARN")
	critCount := flag.Int("crit", 3, "Minimum reports to return a CRIT")
	versionPrint := flag.Bool("version", false, "Prints the version and exits")

	flag.Parse()

	if *versionPrint {
		fmt.Printf("abuseipdb-check - version %s (compiled at %s)\nCopyright (c) 2024 webförsterei Softwareentwicklung GmbH & Co. KG (MIT-License applies)\ngithub.com/webfoersterei/abuseipdb-check\n", version, compileDate)
		os.Exit(ExitCodeOk)
	}

	if len(strings.Trim(*address, " ")) == 0 {
		fmt.Print("Invalid arguments: No address provided via -host")
		os.Exit(ExitCodeUnknown)
	}

	if len(strings.Trim(*apiKey, " ")) == 0 {
		fmt.Print("Invalid arguments: No API-Key provided via -key")
		os.Exit(ExitCodeUnknown)
	}

	if *warnCount > *critCount {
		fmt.Printf("Invalid arguments: WarnCount (%d) greater then CritCount (%d)\n", *warnCount, *critCount)
		os.Exit(ExitCodeUnknown)
	}

	apiResult, err := getEntryCount(apiKey, address, daysToCheck)
	if err != nil {
		fmt.Println("UNKNOWN - Error: ", err)
		os.Exit(ExitCodeUnknown)
	}

	statusMessage := buildStatusMessage(apiResult)

	if apiResult.TotalReports >= *critCount {
		fmt.Printf("CRITICAL - %s", statusMessage)
		os.Exit(ExitCodeCritical)
	} else if apiResult.TotalReports >= *warnCount {
		fmt.Printf("WARNING - %s", statusMessage)
		os.Exit(ExitCodeWarning)
	}

	fmt.Printf("OK - %s", statusMessage)
	os.Exit(ExitCodeOk)
}

func getEntryCount(apiKey *string, address *string, daysToCheck *int) (AbuseIpDbCheckData, error) {
	url := fmt.Sprintf("https://api.abuseipdb.com/api/v2/check?ipAddress=%s&maxAgeInDays=%d&verbose=1", *address, *daysToCheck)

	client := &http.Client{
		Timeout: time.Second * 30,
	}

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return AbuseIpDbCheckData{}, err
	}

	req.Header.Add("User-Agent", fmt.Sprintf("check_abuseipdb/%s (golang/http; Github: webfoersterei)", version))
	req.Header.Add("Accept", "application/json")
	req.Header.Add("Key", *apiKey)
	resp, err := client.Do(req)
	if err != nil {
		return AbuseIpDbCheckData{}, err
	}

	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return AbuseIpDbCheckData{}, err
	}

	if resp.StatusCode != 200 {
		return AbuseIpDbCheckData{}, errors.New(string(bytes.TrimSpace(body)))
	}

	response := AbuseIpDbCheckResponse{}
	err = json.Unmarshal(body, &response)
	if err != nil {
		return AbuseIpDbCheckData{}, err
	}

	return response.Data, err
}
